docs/binary-search-tree/RedBlackTree.md

```diff
@@ -0,0 +1,246 @@
+---
+id: rbt-intro
+sidebar_position: 17
+title: Red-Black Trees
+sidebar_label: Red-Black Trees
+description: "In this blog post, we'll dive into Red-Black Trees, a self-balancing binary search tree known for efficient operations, widely used in real-world applications like Linux kernel and Java's TreeMap."
+tags: [dsa, data structures, red black tree]
+---
+
+## Introduction
+A **Red-Black Tree** is a self-balancing binary search tree where each node has an additional attribute: a color, which can be either red or black. The primary objective of these trees is to maintain balance during insertions and deletions, ensuring efficient data retrieval and manipulation.
+
+## Definition and Structure
+A Red-Black Tree consists of nodes that follow a set of rules:
+- **Color**: Each node is either red or black.
+- **Data**: The value stored in the node.
+- **Left Child**: Reference to the left subtree (smaller values).
+- **Right Child**: Reference to the right subtree (larger values).
+- **Parent**: Reference to the parent node.
+- **Root is always black**: The root node is black, ensuring a stable structure.
+
+## Properties of Red-Black Trees
+A Red-Black Tree have the following properties:
+- **Node Color:** Each node is either red or black.
+- **Root Property:** The root of the tree is always black.
+- **Red Property:** Red nodes cannot have red children (no two consecutive red nodes on any path).
+- **Black Property:** Every path from a node to its descendant null nodes (leaves) has the same number of black nodes.
+- **Leaf Property:** All leaves (NIL nodes) are black.
+
+These properties ensure that the longest path from the root to any leaf is no more than twice as long as the shortest path, maintaining the tree’s balance and efficient performance.
+
+## Rules for constructing Red-Black Tree:
+- The black height of the red-black tree is the number of black nodes on a path from the root node to a leaf node. 
+- Leaf nodes are also counted as black nodes. So, a red-black tree of height h has black height >= h/2.
+- Height of a red-black tree with n nodes is h<= 2 log2(n + 1).
+- All leaves (NIL) are black.
+- The black depth of a node is defined as the number of black nodes from the root to that node i.e the number of black ancestors.
+
+## Correct Red-Black Tree:
+
+    10B
+   /   \
+  5R    20B
+ / \    /  \
+2B  8B 15R 25B
```

Comment:
replace 

```md
    10B
   /   \
  5R    20B
 / \    /  \
2B  8B 15R 25B
```

with

````md
```
    10B
   /   \
  5R    20B
 / \    /  \
2B  8B 15R 25B
```
````

**OR**

if possible then try to use the mermaid graph (https://mermaid.js.org/)